File: api/internal/models/mediaitem.go
package models

import "time"

const ColMediaItems = "mediaitems"

type (
	MediaItem struct {
		ID                string         `json:"id" bson:"_id"`
		Description       string         `json:"description"`
		PreviewURL        string         `json:"previewUrl"`
		SourceURL         string         `json:"sourceUrl"`
		MimeType          string         `json:"mimeType"`
		FileName          string         `json:"fileName"`
		FileSize          int64          `json:"fileSize"`
		MediaMetadata     *MediaMetaData `json:"mediaMetadata"`
		ContentCategories []string       `json:"contentCategories"`
		Entities          []string       `json:"entities"`
		Albums            []string       `json:"albums"`
		Favourite         bool           `json:"favourite"`
		Deleted           bool           `json:"deleted"`
		CreatedAt         time.Time      `json:"createdAt"`
		UpdatedAt         time.Time      `json:"updatedAt"`
		Status            string         `json:"status"`
	}

	MediaMetaData struct {
		CreationTime time.Time `json:"creationTime"`
		Width        *int      `json:"width"`
		Height       *int      `json:"height"`
		Photo        *Photo    `json:"photo"`
		Video        *Video    `json:"video"`
		Location     *Location `json:"location"`
	}

	Photo struct {
		CameraMake      *string  `json:"cameraMake"`
		CameraModel     *string  `json:"cameraModel"`
		FocalLength     *float64 `json:"focalLength"`
		ApertureFNumber *float64 `json:"apertureFNumber"`
		IsoEquivalent   *int     `json:"isoEquivalent"`
		ExposureTime    *float64 `json:"exposureTime"`
	}

	Video struct {
		CameraMake  *string `json:"cameraMake"`
		CameraModel *string `json:"cameraModel"`
		Fps         *int    `json:"fps"`
		Status      *string `json:"status"`
	}

	Location struct {
		Latitude  *float64 `json:"latitude"`
		Longitude *float64 `json:"longitude"`
	}
)

File: tests/helpers/mediaitem.py
import json
from .common import get_response


def get_file_name(ext: str):
    return 'sample.{}'.format(ext.lower())

def upload(file_type):
    file_name = get_file_name(file_type)
    with open(f'data/mediaitem/images/{file_name}', 'rb') as f:
        res = get_response(
            query="""
                mutation Upload($file: Upload!) {
                    upload(file: $file)
                }
            """,
            variables={"file": f},
            upload_files=True
        )
        return res['upload'] if res is not None else None

def get_mediaitem(id):
    res = get_response(
        query="""
            query MediaItem($id: String!) {
                mediaItem(id: $id) {
                    id
                    description
                    fileName
                    mimeType
                    fileSize
                    favourite
                    deleted
                    sourceUrl
                    previewUrl
                    mediaMetadata {
                        width
                        height
                        photo {
                            cameraMake
                            cameraModel
                            focalLength
                            apertureFNumber
                            isoEquivalent
                            exposureTime
                        }
                    }
                }
            }
        """,
        variables={"id": id}
    )
    return res['mediaItem'] if res is not None else None

def get_mediaitems():
    res = get_response(
        query="""
            query MediaItems {
                mediaItems {
                    nodes {
                        id
                    }
                }
            }
        """,
        variables={}
    )
    return res['mediaItems']['nodes'] if res is not None else None

def update_mediaitem_description(id, description):
    res = get_response(
        query="""
            mutation UpdateMediaItemDescription($id: String!, $description: String!) {
                updateDescription(id: $id, description: $description)
            }
        """,
        variables={"id": id, "description": description}
    )
    return res['updateDescription'] if res is not None else None

def favourite_mediaitem(id, type):
    res = get_response(
        query="""
            mutation Favourite($id: String!, $type: String!) {
                favourite(id: $id, type: $type)
            }
        """,
        variables={"id": id, "type": type}
    )
    return res['favourite'] if res is not None else None

def get_favourites():
    res = get_response(
        query="""
            query Favourites {
                favourites {
                    nodes {
                        id
                    }
                }
            }
        """,
        variables={}
    )
    return res['favourites']['nodes'] if res is not None else None

def delete_mediaitem(id, type):
    res = get_response(
        query="""
            mutation Delete($id: String!, $type: String!) {
                delete(id: $id, type: $type)
            }
        """,
        variables={"id": id, "type": type}
    )
    return res['delete'] if res is not None else None

def get_deleted():
    res = get_response(
        query="""
            query Deleted {
                deleted {
                    nodes {
                        id
                    }
                }
            }
        """,
        variables={}
    )
    return res['deleted']['nodes'] if res is not None else None

def get_mediaitem_by_file_type(db, mime_type):
    res = db['mediaitems'].find_one({'mimeType': mime_type})
    if res is not None:
        return str(res['_id'])
    return None

def get_expected_metadata(file_type):
    file_name = get_file_name(file_type)
    with open(f'data/mediaitem/metadata/{file_name}.json', 'rb') as f:
        return json.load(f)
